ukvm/ukvm_cpu_aarch64.c

```diff
@@ -0,0 +1,495 @@
+
+    /* Setup Translation Control Register EL1 */
+    data = TCR_EL1_INIT;
+    ret = aarch64_set_one_register(vcpufd, TCR_EL1, data);
+    if (ret == -1)
+         err(1, "KVM: Setup Translation Control Register EL1 failed");
+
+    /* Setup Translation Table Base Register 0 EL1 */
+    data = AARCH64_PAGE_TABLE;
+    ret = aarch64_set_one_register(vcpufd, TTBR0_EL1, data);
+    if (ret == -1)
+         err(1, "KVM: Translation Table Base Register 0 EL1 failed");
+
+    /* Setup Translation Table Base Register 1 EL1 */
+    data = AARCH64_PAGE_TABLE;
+    ret = aarch64_set_one_register(vcpufd, TTBR1_EL1, data);
+    if (ret == -1)
+         err(1, "KVM: Setup Translation Table Base Register 1 EL1 failed");
+
+    /* Setup System Control Register EL1 to enable MMU */
+    ret = aarch64_get_one_register(vcpufd, SCTLR_EL1, &data);
+    if (ret == -1)
+         err(1, "KVM: Get System Control Register EL1 failed");
+
+    /* Enable MMU and I/D Cache for EL1 */
+    data |= (_SCTLR_M | _SCTLR_C | _SCTLR_I);
+    ret = aarch64_set_one_register(vcpufd, SCTLR_EL1, data);
+    if (ret == -1)
+         err(1, "KVM: Setup System Control Register EL1 failed");
+}
+
+int ukvm_aarch64_dump_pc(int vcpufd, uint64_t *pdata)
+{
+    return aarch64_get_one_register(vcpufd, REG_PC, pdata);
+}
+
+void ukvm_aarch64_setup_vectors(int vcpufd, ukvm_gpa_t vbar_addr)
+{
+    int ret;
+
+    ret = aarch64_set_one_register(vcpufd, VBAR_EL1, vbar_addr);
+    if (ret == -1)
+        err(1, "KVM: Setup Vector Base Address Register EL1 failed");
+}
+
+/* Map a userspace memroy range as guest physical memroy. */
+void ukvm_aarch64_setup_memory(int vmfd, void* vaddr,
+                               uint64_t guest_phys_addr, uint64_t size)
+{
+    int ret, idx;
+    uint32_t used_slot = 0;
+    uint64_t elf_va_end = 0, elf_pa_end = 0;
+    struct kvm_userspace_memory_region *elf_region;
+    struct kvm_userspace_memory_region region;
+
+    /* Register boot_info and unused region before ELF regions */
+    region.slot = 0;
+    region.flags = 0;
+
+    /* We will skip first 4K bytes for MMIO */
+    region.guest_phys_addr = guest_phys_addr + AARCH64_MMIO_SZ;
+    region.memory_size = AARCH64_GUEST_MIN_BASE - AARCH64_MMIO_SZ;
+
+    /*
+     * To keep the pa and va have the same offset in elf, we should skip
+     * first 4K bytes in va too.
+     */
+    region.userspace_addr = (uint64_t)(vaddr + AARCH64_MMIO_SZ);
+
+    ret = ioctl(vmfd, KVM_SET_USER_MEMORY_REGION, &region);
+    if (ret == -1)
+        err(1, "KVM: ioctl (SET_USER_MEMORY_REGION) slot=%d failed", used_slot);
+
+    used_slot++;
+    /* Register elf regions */
+    for (idx = 0; idx < ukvm_region_list->count; idx++) {
+        uint64_t va_end, pa_end;
+        elf_region = ukvm_region_list->regions + idx;
+
+        /* Update slot number with used slot before elf regions */
+        elf_region->slot = used_slot;
+        ret = ioctl(vmfd, KVM_SET_USER_MEMORY_REGION, elf_region);
```

Comment:
Hmm, I need to re-consider it and try to make it be simple. It's best to avoid using regions if possible.